---
title: "yapima"
author:
  - name: Xavier Pastor
    affiliation:
    - Heidelberg Center for Personalized Oncology (DKFZ-HIPO), Heidelberg, Germany
    - Division of Theoretical Bioinformatics, German Cancer Research Center (DKFZ), Heidelberg, Germany
    email: xavier.pastor@compbio-dev.com

date: "`r Sys.Date()`"
output:
  github_document:
    output_file: README.md
    toc: true
    toc_float: true
    number_sections: true
link-citations: true
bibliography: REFERENCES.bib
vignette: >
  %\VignetteIndexEntry{CopyNeutralIMA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment='#>')
```
# yapima

## Description
*yapima* is a pipeline developed to simplify and standardize the preprocessing of the Illumina InfiniumHumanMethylation arrays in order to easily produce tables with beta-values and/or M-values that can be later used for further analysis.

For the sake of simplicity in the usage and comprehension of the pipeline most of the options are hardcoded, and the interaction with it almost reduced to input/output definition and processes switches.

## Definition of variables
### Requirements
*yapima* uses tools that rely on the standard Illumina structures to import the data, and although it limits the flexibility of the tool, allows the user to provide the data without much need of manipulation since all the required files are provided by the Core Facility. There are two requirements regarding input data.

* **IDAT_DIR**: the path to the directory with all the idat files. Initially the core facility provides a compressed file (suffix *idat.7z*) for each array with all the idat files for that array. It is necessary to uncompress these files so that in the end all the idat files are located in the same directory  
* **SAMPLE_ANNOTATION**: path to a SampleSheet file
* **OUTDIR**: directory to store all the files generated by *yapima*

#### A few rules for the SampleSheet
The SampleSheet will always be provided by the core facility, one for each array, so in case a project requires multiple arrays the files need to be concatenated and only should contain the samples related to the project. The explanations here are to understand the final structure of the Sample Sheet and besides concatenating the files and adding additional information in general nothing else needs to be done.

The basic SampleSheet is a comma delimited file that always must have the following columns:

* Sample_Name
* Sentrix_ID: array barcode must correspond to its actual value for proper identification of the idat files
* Sentrix_Position: array position (6x2 disposition of the 12 samples) must correspond to the actual position of the sample in the array in order to properly identify the idat files; it must follow the following pattern: R0[1-8]C0[12]

It is mandatory that the columns *Sample_Name*, *Sentrix_ID* and *Sentrix_Position* contain proper values. For the other columns, if no values are provided they must be empty (i.e. no 'NA' values), so the final structure of the file must be something like this:

```
Sample_Name,Sentrix_ID,Sentrix_Position  
Normal1,1234567890,R01C01  
Normal2,1234567890,R02C01  
Normal3,1234567890,R03C01  
Normal4,1234567890,R04C01  
Normal5,1234567890,R05C01  
Normal6,1234567890,R06C01  
Tumor1,1234567890,R01C02  
Tumor2,1234567890,R02C02  
Tumor3,1234567890,R03C02  
Tumor4,1234567890,R04C02  
Tumor5,1234567890,R05C02  
Tumor6,1234567890,R06C02  
```

Optionally, other columns of interest for the differential methylation analysis or the correction of batch effects can be added. In such case, if the value of a sample is unknown 'NA' must be provided. The values of categorical variables can not be exclusively numerical (i.e.: absence/presence of a mutation should be referred as something like 'ref'/'mut', instead of 0/1), although numerical values may be used if other values are in string format. However, this numerical values will lose their numerical properties. Also, columns with numerical variables can include only numerical values or 'NA'.

```
Sample_Name,Sentrix_ID,Sentrix_Position,Group,Hospital  
Normal1,1234567890,R01C01,Normal,AAA  
Normal2,1234567890,R02C01,Normal,AAA  
Normal3,1234567890,R03C01,Normal,BBB  
Normal4,1234567890,R04C01,Normal,NA  
Normal5,1234567890,R05C01,Normal,BBB  
Normal6,1234567890,R06C01,Normal,BBB  
Tumor1,1234567890,R01C02,Tumor,BBB  
Tumor2,1234567890,R02C02,Tumor,BBB  
Tumor3,1234567890,R03C02,Tumor,AAA  
Tumor4,1234567890,R04C02,Tumor,NA  
Tumor5,1234567890,R05C02,Tumor,AAA  
Tumor6,1234567890,R06C02,Tumor,BBB  
```

### Optional variables
* **BLACKLIST**: stores the path to a file with a set of probes that should be discarded from the very beginning of the analysis for whatever reason. The probes should be listed per line and identified by the Illumina identifier that generally starts with *cg*. Each line can have one or two columns, the second one being the sample where the value related to the probe should be masked. In summary, a probe not related to any sample will be flagged but the values will be left. In case a probe is related to a sample, the value for that probe in that sample will be replaced by 'NA'. A probe can be related to multiple samples by adding a line for each sample. The 'BLACKLIST' file can be mixed with both types of probes to be masked, global or sample specific.
* **RUN_CNV**: valid values are T/TRUE or F/FALSE and toggles the execution of CNV analysis
* **RUN_DIFFERENTIAL_METHYLATION**: valid values are T/TRUE or F/FALSE and toggles the execution of differential analysis
* **REMOVE_SNPS**: possible values are T/TRUE or F/FALSE and flags the polymorphic probes of a given population. By default 'TRUE'
* **POPULATION**: population from which the allelic frequency is used to flag polymorphic probes. By default 'EUR'. Possible values are '', 'EUR', 'AMR', 'AFR', ASN' for 450k arrays and '', 'EUR', 'AMR', 'AFR', 'EAS', 'SAS' for EPIC arrays
* **USE_PREDICTED_SEX**: possible values are T/TRUE of F/FALSE, to define if the predicted sex has to be used as confounder
* **BATCH_VARS**: comma separated list of columns, without spaces, from the sample sheet that will be used as confounders or batch effects
* **SEED**: an integer to allow the repetition of results in random processes

## Execution
*yapima* is executed in a very similar way to other pipelines, using a bash configuration file that needs to be sourced:
```
sh /path/to/yapima/yapima.sh -c /path/to/config_yapima.sh
```

## Analysis steps
Since the pipeline is implemented in R and the config file is just a bash front-end for the pipeline, the values passed to the variables that control the processess switching need valid R logical values or expressions that can be evaluated in R context, not in bash. However, to make it easier, it's also possible to switch off/on the processes with 0/1.

### Preprocessing
The data is loaded into the R environment using the *minfi* [@minfi] package from bioconductor. Initially, probes are flagged using the following values and summed accordingly, being 0 the value for probes that pass all the filters:
* 8: probes with low quality in at least half of the samples in a cohort
* 4: probes provided in **BLACKLIST**
* 2: crossreactive probes for the IlluminaHumanMethylation450k array [@450k], or for the IlluminaHumanMethylationEPIC array [@epic]
* 1: if selected, probes with SNPs at the CpG or SBE position that with an allele frequency of 0.005 in the selected population for the IlluminaHumanMethylation450k array [@450k] or for the IlluminaHumanMethylationEPIC array [@epic]

Afterwards, *ENmix* 'est' background correction [@ENmix] and 'RELIC' dye bias correction [@RELIC] are performed, and RCP normalization [@RCP] is applied. Measures with low detection are replaced by 'NA'.

### Copy Number Variation
The CNV analysis is done using the *conumee* [@conumee] package from bioconductor. This is one of the longest parts of the pipeline. In general, for 20 samples the basic execution of the pipeline (normalization, batch correction and differential methylation analysis) may take around 10 minutes, but the CNV analysis may need other 45 minutes, so when executing the pipeline for exploratory purposes it may be convenient to switch off this part.

*conumee* requires a sample set without chromosomal aberrations that are provided by *CopyNeutralIMA* [@cnima].

This analysis may provide a good idea about chromosomal aberrations although it shouldn't replace an analysis at genomic level, and it proofs to be as valid to detect sample swaps as CNV analysis using sequencing data.

### Differential Methylation Analysis
The analysis of differentially methylated positions (DMP) is done using the *[limma](https://bioconductor.org/packages/release/bioc/html/limma.html)* [@limma] package from bioconductor and following the procedures detailed by the authors. It can perform only paired and unpaired analysis and reports tables with P-values and adjusted P-values (Benjamini & Hochberg [@fdr]), and takes as variables of interest all the extra columns of the SampleSheet except the provided in the **BATCH_VARS** variable, although these are also included in the linear model to account for possible confounding factors.

The detection of differentially methylated regions (DMR) is done using the *[DMRcate](https://www.bioconductor.org/packages/release/bioc/html/DMRcate.html)* [@dmr] package from bioconductor. For two groups comparisons, the t statistics from the DMP analysis are used, and the beta log fold change is computed running the standard limma workflow on the beta values. For comparisons with more than two groups the squared F statistics from the DMP analysis are provided and the beta log fold change is set to 0. All the other parameters are left as default.

## Output
### Basic output
When the pipeline is run without switching on any of the steps the following files are produced:

* **annotation.txt**: tab delimited file with annotation for each probe in InfiniumHumanMethylation450k arrays
* **filtered_raw_betas.gz**: compressed BED file with beta values before any transformation
* **genotyping_betas.txt**: matrix of beta values of 65 genotyping probes present in the array
* **filtered_normalized_meth.RData**: an object of class *MethylSet* produced after transforming data with *ENmix* background correction and *RCP* normalization
* **filtered_normalized_betas.gz**: a compressed BED file with transformed beta-values, ready for analysis
* **filtered_normalized_M.gz**: compressed BED file with transformed M-values, suited for statistical analysis
* **extended_sample_sheet.csv**: SampleSheet with an additional column for the predicted gender of the samples

#### Reproducible research
The pipeline also produces two files in order to be able to reproduce the analysis.
* **methods_session_references.docx**: contains a description of the analysis as a paragraph that could be added to the *Methods* section of a paper; the session info with all the packages and versions used to reproduce the environment used for the analysis and the commit id for the version of the pipeline executed; the references to be mentioned in case the analysis is included in a paper
* **script.R**: R script that reproduces a transparent analysis to repeat the results produced by the pipeline, with all the conditions evaluated by the pipeline already removed

#### QC
By default, the following QC files are also produced:

* **qc/beta_distribution.pdf**: density plots of beta values for each sample, befor and after transformation
* **qc/beta_distribution_heatmap.pdf**: density heatmaps of beta values, before and after transformation; these plots help better to identify the samples with abnormal beta values distribution
* **qc/raw_batch_PCA_***: PCA plots of samples before transformation identifying the samples by groups of batch variables
* **qc/processed_batch_PCA_***: PCA plots of samples after transformation identifying the samples by groups of batch variables
* **qc/samples_correlation.pdf**: heatmaps showing sample correlations using beta values after transformation
* **qc/median_intensities.pdf**: scatter plot showing the median intensity for the methylated vs unmethylated intensities in each sample
* **qc/bisulfite_conversion.pdf**: dotplots showing the success of the bisulfite conversion
* **qc/genotypes.txt**: BED file with inferred genotypes of the samples using the beta values of genotyping probes

Additionally, if there are columns in the SampleSheet other than the basic ones or identified as batch effects the following plots are produced:

* **qc/raw_PCA_***: PCA plots of samples before transformation identifying the samples by variables of interest
* **qc/processed_PCA_***: PCA plots of samples after transformation identifying the samples by variables of interest

### Copy Number Variation
This is mostly a QC step that produces two files per sample:

* **qc/CNV_report/$pid_CNV_report.pdf**: pdf file with one plot showing all the aberrations in the genome and one plot per chromosome, showing the aberrations present in each chromosome in more detail
* **qc/CNV_report/$pid_CNV_report.txt**: tab delimited file with information on the regions of the genome with chromosomal aberrations

### Differential Methylation Analysis
Each column in the SampleSheet that does not correspond to a basic column or a batch variables is used to define a differential methylation analysis, and for each of them the following files are produced:

* **DMP/DMP_?.bed.gz**: compressed BED file with some statistical parameters (*limma*) like the log fold change ('logFC'), average methylation ('AveExpr'), the P-value ('P.value') and the adjusted P-value('adj.P.Value'), together with array annotation
* **DMP/DMP_PCA_?.png**: PCA plot using the probes with an adjusted P-value smaller than 0.05
* **DMP/DMP_cluster_?.pdf**: hierarchical clustering of the samples using the probes with and adjusted P-value <= 0.05
* **DMR/DMR_?.bed**: BED file with the identified DMRs
* **DMR/DMR_heatmap_?.pdf**: heatmap of CpGs from the top DMRs that sum up to 500 CpGs, being the top DMRs those with a smaller Stouffer.

### Dependencies
The current version of yapima requires R-3.5.0 and Bioconductor 3.8. The following packages and their dependencies must be installed:

* biomaRt
* cluster
* ComplexHeatmap
* conumee
* CopyNeutralIMA
* DMRcate
* ENmix
* fpc
* GenomicRanges
* ggplot2
* grid
* gridExtra
* limma
* minfi
* parallel
* pbapply
* plyr
* tools
* XML
* git2r
* rmarkup
* knitr

Additionally, pandoc v2.2.1 or higher is necessary in order to produce the documentation of the execution.

___
This pipeline has been developed by Xavier Pastor (<xavier.pastor@compbio-dev.com>) at DKFZ, Heidelberg, Germany.

## References
